src/main/java/net/neoforged/neoforge/transfer/storage/StorageUtil.java

```diff
@@ -0,0 +1,370 @@
+            int size = storage.size();
+
+            for (int i = 0; i < size; ++i) {
+                if (!storage.isResourceBlank(i) && storage.getAmount(i) > 0) {
+                    amount += storage.insert(i, resource, maxAmount - amount, transaction);
+                    if (amount == maxAmount) return amount;
+                }
+            }
+
+            for (int i = 0; i < size; ++i) {
+                amount += storage.insert(resource, maxAmount - amount, transaction);
+                if (amount == maxAmount) return amount;
+            }
+        } catch (Exception e) {
+            CrashReport report = CrashReport.forThrowable(e, "Inserting resources into slots");
+            report.addCategory("Insertion details")
+                    .setDetail("Storage", storage::toString)
+                    .setDetail("Resource", () -> Objects.toString(resource, null))
+                    .setDetail("Max amount", maxAmount)
+                    .setDetail("Transaction", transaction);
+            throw new ReportedException(report);
+        }
+
+        return amount;
+    }
+
+    /**
+     * Attempt to find a resource stored in the passed storage.
+     *
+     * @see #findStoredResource(Storage, Predicate)
+     * @return A non-blank resource stored in the storage, or {@code null} if none could be found.
+     */
+    @Nullable
+    public static <T> T findStoredResource(@Nullable Storage<T> storage) {
+        return findStoredResource(storage, r -> true);
+    }
+
+    /**
+     * Attempt to find a resource stored in the passed storage that matches the passed filter.
+     *
+     * @param storage The storage to inspect, may be null.
+     * @param filter The filter. Only a resource for which this filter returns {@code true} will be returned.
+     * @param <T> The type of the stored resources.
+     * @return A non-blank resource stored in the storage that matches the filter, or {@code null} if none could be found.
+     */
+    @Nullable
+    public static <T> T findStoredResource(@Nullable Storage<T> storage, Predicate<T> filter) {
+        Objects.requireNonNull(filter, "Filter may not be null");
+        if (storage == null) return null;
+
+        int slots = storage.size();
+        for (int slot = 0; slot < slots; ++slot) {
+            T resource = storage.getResource(slot);
+            if (filter.test(resource)) {
+                return resource;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Attempt to find a resource stored in the passed storage that can be extracted.
+     *
+     * @see #findExtractableResource(Storage, Predicate, TransactionContext)
+     * @return A non-blank resource stored in the storage that can be extracted, or {@code null} if none could be found.
+     */
+    @Nullable
+    public static <T> T findExtractableResource(@Nullable Storage<T> storage, @Nullable TransactionContext transaction) {
+        return findExtractableResource(storage, r -> true, transaction);
```

Comment:
again, not really a useful predicate as there's no empty check. If a handler is stupidly enough implemented, you'd just get X of AIR

Reply:
Fixed.